services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/client/pulsar/DataPlaneClientImpl.java

```diff
@@ -18,24 +18,151 @@
 import com.futurewei.alcor.dataplane.client.DataPlaneClient;
 import com.futurewei.alcor.dataplane.entity.MulticastGoalState;
 import com.futurewei.alcor.dataplane.entity.UnicastGoalState;
+import com.futurewei.alcor.dataplane.exception.GroupTopicNotFound;
+import com.futurewei.alcor.dataplane.exception.MulticastTopicNotFound;
 import com.futurewei.alcor.schema.Goalstate;
 import com.futurewei.alcor.schema.Goalstateprovisioner;
+import com.futurewei.alcor.web.entity.dataplane.MulticastGoalStateByte;
+import com.futurewei.alcor.web.entity.dataplane.UnicastGoalStateByte;
+
+import org.apache.pulsar.client.api.Producer;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.impl.schema.JSONSchema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
 
+import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
 //@Component
```

Comment:
I remember we had an old issue in the new DPM when we enable gRPC client and pulsar client, DPM can't launch. Not sure if it has been fixed. @chenpiaoping @VanderChen

Reply:
I don't know whether pulsar and gRPC should be enabled at the same time. If not, we plan to use a configuration file to decide which client.

Reply:
Configuration is one option which means that gRRC and Pulsar is exclusive to each other. 

@VanderChen, can you check how to allow two client coexist? Thank you.

Reply:
In the current implementation, there can only be one dataClient instance. Do we need two dataClient instances in one dpm at the same time? if so, which one should be used?

Reply:
The short answer is yes, we would love to need two clients coexist in one DPM instance. The gPRC client is for our fast path, and the pulsar or other MQ clients is for scaling path. 

On the service layer, we could develop an algorithm to decide path to go. A simplest algorithm is purely based on the is_fast field in PortEntity but we could generate more sophisticated algorithm that consider the size of VPC (for the port to be updated), scenarios ( port/sg/route/nacl update), and other parameters.